pkg/stats/statser_prometheus.go

```diff
@@ -0,0 +1,78 @@
+package stats
+
+import (
+	"context"
+	"time"
+
+	"github.com/prometheus/client_golang/prometheus"
+
+	"github.com/atlassian/gostatsd"
+)
+
+// NOTE: is using a collector/vec of gauges/counters reasonable? Or does
+// using only a single gauge/counter suffice?
+
+// this depends on the metrics being collected, if they are the same thing
+// but can be partitioned into different types/groups, e.g. http requests partitioned
+// by user age and demographics, then using a collector makes sense. (is this correct?)
+
+// PrometheusStatser is a Statser that monitors gostasd's internal metrics from
+// Prometheus, it is useful when there is a large number of ephemeral hosts.
+type PrometheusStatser struct {
+	flushNotifier
+
+	// collector of gauges that stores the internal gauge metrics of gostatsd
+	gaugeVec prometheus.GaugeVec
+	// collector of counters that stores the internal count metrics of gostatsd
+	counterVec prometheus.CounterVec
+}
+
+// NewPrometheusStatser creates a new Statser which
+// sends internal metrics to prometheus
+func NewPrometheusStatser(gaugeVec prometheus.GaugeVec, counterVec prometheus.CounterVec) *PrometheusStatser {
+	return &PrometheusStatser{
+		gaugeVec:   gaugeVec,
+		counterVec: counterVec,
+	}
+}
+
+func (ps *PrometheusStatser) NotifyFlush(ctx context.Context, d time.Duration) {
```

Comment:
You don't need to implement `NotifyFlush` and `RegisterFlush`, it's picked up implicitly from the embedded `flushNotifier`

Reply:
How come the `TaggedStatser` implements those two functions or why does `TaggedStatser` not use a `flushNotifier` as one of its attributes?

Reply:
The `TaggedStatser` is a wrapper that explicitly calls to the wrapped `Statser`.

It shouldn't use a `flushNotifier` directly, because then the `TaggedStatser` and the underlying `Statser` would have different notification targets (and ultimately `MetricFlusher.Run` would only call `NotifyFlush` on the underlying `Statser`)

That being said, it could and probably should use an embedded `Statser` and remove the explicit implementation.

Reply:
My bad, that comments from my personal account.

The `TaggedStatser` is a wrapper that explicitly calls to the wrapped `Statser`.

It shouldn't use a `flushNotifier` directly, because then the `TaggedStatser` and the underlying `Statser` would have different notification targets (and ultimately `MetricFlusher.Run` would only call `NotifyFlush` on the underlying `Statser`)

That being said, it could and probably should use an embedded `Statser` and remove the explicit implementation.

Reply:
Oh, I see that now, what was the design process behind, is it because that they have different notification targets, so one such as `TaggedStatser` is a wrapper of  `Statser` but `InternalStatser` is not a wrapper because they have the same notification targets?

I am using the GoLand IDE, and I am trying to find the usage of `NotifyFlush`, for some reason I cannot see any call to this method outside of the `statsers` and the `flush_notifer`.

Reply:
There was no special design considerations behind `TaggedStatser` not embedding the `Statser`, it's possible at the time that I thought you could only embed a struct in a struct, and an interface in an interface, but not an interface in a struct, I don't recall for sure though.

`InternalStatser` does explicitly implement `NotifyFlush` however, as it needs to take action on the flush **before** returning.  Flush notification is an asynchronous, best effort action, and any metrics triggered by it are processed in the next flush cycle.

The `InternalStatser` however explicitly needs to process the metrics in the current flush cycle, so it takes action (dispatching internal metrics) before returning.

I'm not sure why GoLand doesn't detect usage - it does in my setup.  You could try rebuilding the index through File -> Invalidatae Caches / Restart, but I don't know if it will help.

The only place it's called is at https://github.com/atlassian/gostatsd/blob/master/pkg/statsd/flusher.go#L69

Reply:
My problem was I was not aware that the `Statser` interface also has a `NotifyFlush` method, I was only trying to find the usage of `NotifyFlush` from the `flsuhNotifier` struct.

Reply:
Ahhh, that would do it :)